### PR TITLE
Move CWallet::setKeyPool to private section of CWallet.

### PR DESCRIPTION
**Pull request by pstratem**
Further encapsulation.

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -1439,7 +1439,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     LogPrintf("mapBlockIndex.size() = %u\n",   mapBlockIndex.size());
     LogPrintf("nBestHeight = %d\n",                   chainActive.Height());
 #ifdef ENABLE_WALLET
-    LogPrintf("setKeyPool.size() = %u\n",      pwalletMain ? pwalletMain->setKeyPool.size() : 0);
+    LogPrintf("setKeyPool.size() = %u\n",      pwalletMain ? pwalletMain->GetKeyPool().size() : 0);
```
Comment (jonasschnelli) [inline src/init.cpp]: does copy the `std::set<int64_t>` but no keys are copied around. Seems to be okay.

**Comment by jonasschnelli** [comment]
utACK 674b2dec4a9a1ebbdb03d6eac0f4c858ca8793ba

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -594,7 +596,23 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
 
     std::string strWalletFile;
 
-    std::set<int64_t> setKeyPool;
+    std::set<int64_t> GetKeyPool()
+    {
+        return setKeyPool;
+    }
+
+    void LoadKeyPool(int nIndex, CKeyPool &keypool)
```
Comment (paveljanik) [inline src/wallet/wallet.h]: some `const`?

File: src/init.cpp (hunk 1 above)
Comment (sipa) [inline src/init.cpp]: This will copy the entire set. Perhaps a GetKeyPoolSize()?

File: src/wallet/wallet.h (hunk 2 above)
Comment (phantomcircuit) [inline src/wallet/wallet.h]: Indeed this should be const

**Comment by pstratem** [comment]
nits picked

**Comment by MarcoFalke** [comment]
Fails with 

```
Assertion failed: lock cs_wallet not held in ../../src/wallet/wallet.h:849; locks held:

**Comment by phantomcircuit** [comment]
Indeed it does, the logic in master should be acquiring that lock but is
not...

I will correct this oversight.

**Comment by pstratem** [comment]
@MarcoFalke I've fixed that (indeed we were accessing the private class members without the proper lock before).

Ignore the travis error that's a travis error but a pr error.

File: src/init.cpp [hunk 3]
```diff
@@ -1439,9 +1439,12 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     LogPrintf("mapBlockIndex.size() = %u\n",   mapBlockIndex.size());
     LogPrintf("nBestHeight = %d\n",                   chainActive.Height());
 #ifdef ENABLE_WALLET
-    LogPrintf("setKeyPool.size() = %u\n",      pwalletMain ? pwalletMain->setKeyPool.size() : 0);
-    LogPrintf("mapWallet.size() = %u\n",       pwalletMain ? pwalletMain->mapWallet.size() : 0);
-    LogPrintf("mapAddressBook.size() = %u\n",  pwalletMain ? pwalletMain->mapAddressBook.size() : 0);
+    if (pwalletMain) {
```
Comment (MarcoFalke) [inline src/init.cpp]: Nit: `if (!fDisableWallet)` works as well. No strong opinion.

Weak NACK: The goal was to move them into the wallet module. See #7965

**Comment by MarcoFalke** [comment]
Concept ACK e86eb71604e73ad35b8b1f59d73af22e353a156e

**Comment by pstratem** [comment]
@MarcoFalke This should mostly address your concern.

**Comment by MarcoFalke** [comment]
Missing lock:

```
Assertion failed: lock cs_wallet not held in ../../src/wallet/wallet.h:849; locks held:

**Comment by pstratem** [comment]
@MarcoFalke something weird and broken was happening there, I've reduced the scope in which the lock is held and now it works fine...

**Comment by MarcoFalke** [comment]
utACK 8680d3aa800aa4fab8c321b92cc7f3fa97a8a0a3

**Comment by sipa** [comment]
utACK 8680d3aa800aa4fab8c321b92cc7f3fa97a8a0a3

**Comment by jonasschnelli** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/8445/commits/8680d3aa800aa4fab8c321b92cc7f3fa97a8a0a3

**Pull request merged**